flake.nix

```diff
@@ -0,0 +1,160 @@
+        };
+
+        fenix' = fenix.packages.${system};
+
+        toolchainSpec = {
+          channel = "nightly";
+          date = "2022-12-22";
+          sha256 = "sha256-FK01QQuXkFXuy/W7wzAA0G+T2s9dQIDBjMxMC0cUk2M=";
+        };
+
+        platformToTarget = x: {
+          "aarch64-darwin" = "aarch64-apple-darwin";
+          "aarch64-linux" = "aarch64-unknown-linux-gnu";
+          "i686-linux" = "i686-unknown-linux-gnu";
+          "x86_64-darwin" = "x86_64-apple-darwin";
+          "x86_64-linux" = "x86_64-unknown-linux-gnu";
+        }.${x};
+
+        supportedTargets = [
+          "x86_64-unknown-linux-gnu"
+          "x86_64-unknown-linux-musl"
+          "x86_64-apple-darwin"
+          "aarch64-unknown-linux-gnu"
+          "aarch64-unknown-linux-musl"
+          "aarch64-apple-darwin"
+        ];
+
+        allBinaries = [
+          "iroha"
+          "iroha_client_cli"
+          "kagami"
+          "kura_inspector"
+          "parity_scale_decoder"
+        ];
+
+        # Build an Iroha derivation
+        mkIroha =
+          { target ? (platformToTarget system) # target arch to build for
+          , binaries ? allBinaries # which binary to build
+          , name ? "iroha" # resulting derivation name
+          , ...
+          }@args:
+          let
+            toolchainHost = fenix'.toolchainOf toolchainSpec;
+            toolchainTarget = fenix'.targets.${target}.toolchainOf toolchainSpec;
+            toolchain = fenix'.combine [
+              toolchainHost.rustc
+              toolchainHost.cargo
+              toolchainHost.rustfmt
+              toolchainTarget.rust-std
+            ];
+            pkgsCross = (import nixpkgs) {
+              localSystem = system;
+              crossSystem = target;
+              config.allowUnsupportedSystem = true;
+            };
+            naersk = pkgsCross.buildPackages.callPackage inputs.naersk {
+              cargo = toolchain;
+              rustc = toolchain;
+            };
+          in
+          naersk.buildPackage rec {
+            pname = name;
+            version = (builtins.fromTOML (builtins.readFile ./Cargo.toml)).workspace.package.version;
+
+            src = ./.;
+
+            # FIXME: Fails due to https://github.com/rust-lang/cargo/issues/10368
+            # Either try a workaround, or wait for resolution
+            # doDoc = true;
+            # doDocFail = true;
+
+            nativeBuildInputs = with pkgsCross.buildPackages; [ pkg-config libiconvReal gcc binutils ];
+            buildInputs = with pkgsCross; [ openssl.dev libiconvReal zlib ];
+
+            targets = binaries;
+
+            cargoBuildOptions = default:
+              default ++
+              [ "--target" target ] ++
+              builtins.concatMap (binary: [ "-p" binary ]) binaries;
+
+            CARGO_BUILD_TARGET = target;
+
+            CC = "${pkgsCross.buildPackages.gcc}/bin/${target}-cc";
+            TARGET_CC = CC;
+            RUSTFLAGS = "-C linker=${CC}";
```

Comment:
So we're using gnu LD. This is fine for now, but I'd recommend using a faster linker at some point. `mold` and `LD` are on speed parity on Linux, but not on Mac OS

Reply:
I just took what needed least tweaking for now, I'll definitely try to switch it to mold later.